File: uva/tomo103/ArcticNetwork.java
package uva.tomo103;
//Uva 10369.
import java.awt.Point;
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileReader;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;

import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.Locale;
import java.util.StringTokenizer;

public class ArcticNetwork {
	static BufferedReader input;

	static StringTokenizer _stk;

	static String readln() throws IOException {
		String l = input.readLine();
		if (l != null)
			_stk = new StringTokenizer(l, " ");
		return l;
	}

	static String next() {
		return _stk.nextToken();
	}

	static int nextInt() {
		return Integer.parseInt(next());
	}

	static void dbg(Object... o) {
		System.out.println(Arrays.deepToString(o));
	}
	
	static PrintWriter output = new PrintWriter(new BufferedWriter(
			new OutputStreamWriter(System.out)));

	public static void main(String[] args) throws IOException {
		Locale.setDefault(Locale.US);
		input = new BufferedReader(new InputStreamReader(System.in));
		input = new BufferedReader(new FileReader("ArcticNetwork"));
		readln();
		int testCases = nextInt();
		for (int i = 0; i < testCases; i++) {
			readln();
			S = nextInt();
			V = nextInt();
			outpost.clear();
			map.clear();
			EdgeList.clear();
			for (int j = 0; j < V; j++) {
				readln();
				outpost.add(new Point(nextInt(), nextInt()));
			}

			annadirVertices();
			kruskal();
		}
		output.close();
	}

	static double dist;

	private static void annadirVertices() {
		for (int i = 0; i < V; i++) {
			for (int j = i + 1; j < V; j++) {
				u = mappear(outpost.get(i));
				v = mappear(outpost.get(j));
				dist = outpost.get(i).distance(outpost.get(j));
				EdgeList.add(new EdgeL(u, v, dist));
			}
		}
	}

	private static int mappear(Point point) {
		Integer idx = map.get(point);
		if (idx == null) {
			idx = map.size();
			map.put(point, idx);
		}
		return idx;
	}

	static ArrayList<Point> outpost = new ArrayList<Point>(501);
	static HashMap<Point, Integer> map = new HashMap<Point, Integer>(501);
	// KRUSKALL
	static int V, E, u, v, w;
	static int S;
	static int MAX_VERTEX = 501;// depende del problema
	static ArrayList<EdgeL> EdgeList = new ArrayList<EdgeL>(500 * 500);
	static int mst_cost = 0;

	private static void kruskal() {
		Collections.sort(EdgeList); // sort by edge weight in O(E log E)

		double max = 0.0;
		initSet(V); // all V are disjoint sets initially
		for (int i = 0; i < EdgeList.size(); i++) { // tambien se puede
													// conecciones < vertex-1
			if (_numDisjointSets <= S) {
				break;
			}
			// y meter conecciones++ en el if
			EdgeL front = EdgeList.get(i);
			if (!isSameSet(front.from, front.to)) { // if no cycle
				max = Math.max(max, front.cost);
				unionSet(front.from, front.to); // link endpoints
			}
		} // note: the runtime cost of UFDS is very light
		output.printf("%.2f\n", max);

	}

	// Union-Find
	static int[] pset, setSize, rank;
	static int _numDisjointSets;

	static void initSet(int N) {
		rank = new int[N];
		setSize = new int[N];
		Arrays.fill(setSize, 1);
		_numDisjointSets = N;
		pset = new int[N];
		for (int i = 0; i < N; i++) {
			pset[i] = i;
		}
	}

	static int findSet(int i) {
		return (pset[i] == i) ? i : (pset[i] = findSet(pset[i]));
	}

	static boolean isSameSet(int i, int j) {
		return findSet(i) == findSet(j);
	}

	static void unionSet(int i, int j) {
		int xroot = findSet(i), yroot = findSet(j);
		if (xroot == yroot)
			return;
		_numDisjointSets--;
		if (rank[xroot] < rank[yroot]) {
			pset[xroot] = yroot;
			setSize[yroot] += setSize[xroot];
		} else if (rank[xroot] > rank[yroot]) {
			pset[yroot] = xroot;
			setSize[xroot] += setSize[yroot];
		} else {
			pset[yroot] = xroot;
			rank[xroot]++;
			setSize[xroot] += setSize[yroot];
		}
	}

	static int sizeOfSet(int i) {
		return setSize[findSet(i)];
	}
}

class EdgeL implements Comparable<EdgeL> {
	int from, to;
	Double cost;

	public int compareTo(EdgeL e) {
		return cost.compareTo(e.cost);
	}

	public EdgeL(int from, int to, Double cost) {
		this.from = from;
		this.to = to;
		this.cost = cost;
	}

	@Override
	public String toString() {
		return "(from=" + from + ", to=" + to + ", cost=" + cost + ")";
	}
}

File: uva/tomo103/ArcticNetwork.cpp
//Uva 10369.
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <cmath>
#include <ctime>
#include <cctype>
#include <cassert>
#include <iostream>
#include <sstream>
#include <iomanip>
#include <string>
#include <vector>
#include <stack>
#include <queue>
#include <deque>
#include <list>
#include <set>
#include <map>
#include <bitset>
#include <algorithm>
#include <numeric>
#include <complex>

#define D(x) cout << #x << " = " << (x) << endl;
#define REP(i,a,n) for(int i=(a); i<(int)(n); i++)
#define FOREACH(it,v) for(__typeof((v).begin()) it=(v).begin(); it!=(v).end(); ++it)
#define ALL(v) (v).begin(), (v).end()

using namespace std;

typedef long long int64;

const int INF = (int)(1e9);
const int64 INFLL = (int64)(1e18);
const double EPS = 1e-13;


typedef pair<int, int> ii;
typedef vector<int> vi;
typedef vector<ii> vii;

// Union-Find Disjoint Sets Library
// PS: This library code does not include the `union by rank' heuristic yet
// 1000 is just a rough number, adjustable by user
vi pset(1000), setSize(1000); int _numDisjointSets;
void initSet(int N) { setSize.assign(N, 1); _numDisjointSets = N; 
pset.assign(N, 0); for (int i = 0; i < N; i++) pset[i] = i; }
int findSet(int i) { return (pset[i] == i) ? i : (pset[i] = findSet(pset[i])); }
bool isSameSet(int i, int j) { return findSet(i) == findSet(j); }
void unionSet(int i, int j) { 
  if (!isSameSet(i, j)) { 
    _numDisjointSets--; 
    setSize[findSet(j)] += setSize[findSet(i)]; 
    pset[findSet(i)] = findSet(j); } }
int numDisjointSets() { return _numDisjointSets; }
int sizeOfSet(int i) { return setSize[findSet(i)]; }

vector<vii> AdjList;
vi taken;                                  // global boolean flag to avoid cycle
priority_queue<ii> pq;            // priority queue to help choose shorter edges
map<ii, int> mapp;
vii ou;
vector< pair<double, ii> > EdgeList;
int V, E, u, v, S, tc;
    
int mappear(ii point) {
    int idx;
	if (!mapp.count(point)) {
		idx = mapp.size();
		mapp[point]=idx;
	}
	return mapp[point];
}

double dist(ii p1, ii p2) { // Euclidean distance
  // hypot(dx, dy) returns sqrt(dx * dx + dy * dy)
    return hypot((double)p1.first - (double)p2.first, (double)p1.second - (double)p2.second); 
} // return double

void annadirVertices(){
    int u,v;
    double d;
    REP(i,0,V){
        REP(j,0,V){
            u=mappear(ou[i]);
            v=mappear(ou[j]);
            d= dist(ou[i],ou[j]);
            EdgeList.push_back(make_pair(d, ii(u, v)));
        }
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    
//    freopen("ArcticNetwork.txt", "r", stdin);
//    freopen("ArcticNetwork_out.txt", "w", stdout);

    scanf("%d", &tc);
    REP(idCases,0,tc){
        ou.clear();
        mapp.clear();
        scanf("%d %d", &S, &V);
        REP(i,0,V){
            scanf("%d %d", &u, &v);   
            ou.push_back(ii(u, v));
        }
        EdgeList.clear();
        annadirVertices();
        
        sort(EdgeList.begin(), EdgeList.end());   // sort by edge weight in O(E log E)

        double mx = 0.0; initSet(V);             // all V are disjoint sets initially
        for (int i = 0; i < EdgeList.size(); i++) {
            if(_numDisjointSets<=S){
                break;
            }
            
            pair<double, ii> front = EdgeList[i];
            if (!isSameSet(front.second.first, front.second.second)) {    // if no cycle
                mx = max(mx, front.first);
                unionSet(front.second.first, front.second.second);
            } 
        }

        printf("%.2f\n", mx);
    }
}

